spec/index.bs

```diff
@@ -4625,16 +4664,46 @@ interface mixin GPUProgrammablePassEncoder {
-                dynamicOffsetsDataLength:
+                |index|: The index to set the bind group at.
+                |bindGroup|: Bind group to use for subsequent render or compute commands.
+                |dynamicOffsetsData|: Array containing buffer offsets in bytes for each entry in
+                    |bindGroup| with marked as {{GPUBindGroupLayoutEntry/hasDynamicOffset}}.
+                |dynamicOffsetsDataStart|: Offset in elements into |dynamicOffsetsData| where the
+                    buffer offset data begins.
+                |dynamicOffsetsDataLength|: Number of buffer offsets to read from |dynamicOffsetsData|.
             </pre>
 
             **Returns:** void
 
-            Issue: Describe {{GPUProgrammablePassEncoder/setBindGroup(index, bindGroup, dynamicOffsetsData, dynamicOffsetsDataStart, dynamicOffsetsDataLength)}} algorithm steps.
+            Issue the following steps on the [=Device timeline=] of |this|.{{GPUObjectBase/[[device]]}}:
+            <div class=device-timeline>
+                1. If any of the following conditions are unsatisfied, make |this| [=invalid=] and stop.
+                    <div class=validusage>
+                        - |bindGroup| is [=valid=].
+                        - |bindGroup|.{{GPUObjectBase/[[device]]}} is |this|.{{GPUObjectBase/[[device]]}}.
+                        - |index| &lt; {{GPULimits/maxBindGroups|GPULimits.maxBindGroups}}.
+                        - |dynamicOffsetsDataLength| is
+                            |bindGroup|.{{GPUBindGroup/[[layout]]}}.{{GPUBindGroupLayout/[[dynamicOffsetCount]]}}.
+                        - |dynamicOffsetsDataStart| + |dynamicOffsetsDataLength| &le; |dynamicOffsetsData|.length.
+
+                        - Let |dynamicOffsetIndex| be |dynamicOffsetsDataStart|.
+                        - For each {{GPUBindGroupLayoutEntry}} |bindingDescriptor| in
+                            |bindGroup|.{{GPUBindGroup/[[layout]]}}.{{[[entryMap]]}}:
+
+                            - If |bindingDescriptor|.{{GPUBindGroupLayoutEntry/hasDynamicOffset}} is `true`:
+
+                                - Let |bufferBinding| be the {{GPUBindGroupEntry/resource}} of the
+                                    {{GPUBindGroupEntry}} in |bindGroup|.{{GPUBindGroup/[[entries]]}}
+                                    that corresponds with |bindingDescriptor|.
+                                - Let |bufferDynamicOffset| be |dynamicOffsetsData|[|dynamicOffsetIndex|].
+                                - Set |dynamicOffsetIndex| to be |dynamicOffsetIndex| + `1`
+                                - |bufferBinding|.{{GPUBufferBinding/offset}} + |bufferDynamicOffset| +
+                                    |bindingDescriptor|.{{GPUBindGroupLayoutEntry/minBufferBindingSize}}
+                                    &le; |bufferBinding|.{{GPUBufferBinding/buffer}}.{{GPUBuffer/[[size]]}}.
+                                - Let |dynamicOffsetIndex| be |dynamicOffsetIndex| + `1`
```

Comment:
can we move this algorithm of walking the dynamic indices out, so that it is reused in multiple places?

Reply:
Took a stab at this, not sure if the way I structured it is the best route. Thoughts?